### ISSUE TITLE
[feature request]: support mode='raw' in jnp.linalg.qr

### ISSUE THREAD

**Issue by alewis** [google/jax#2322]
The LAPACK implementation of the QR decomposition does not construct the matrices `Q` and `R` directly, but rather lists of (the essential parts of) 'Householder vectors' `h` and scale factors `tau` from which these can be computed straightforwardly; `Q` and `R` are then computed from these via a second LAPACK call.

By default `(o)np.linalg.qr` bundles these calls together and just returns `Q` and `R`. However, when an `m` by `n` matrix with `n` << `m` has been decomposed, working directly with the `h, tau` representation can yield a major performance advantage, since applying `Q` via it scales like m x n rather than m^2 in the dense case. 

Numpy therefore provides a `mode='raw'` option to `np.linalg.qr` that returns `h, tau` (in column major order, but it's still better than nothing). Unfortunately Jax instead raises `NotImplemented`. Note this appears not to be documented.

Digging through the source code, the CPU and GPU implementations actually do retrieve `h` and `tau`, so in these cases supporting `mode=raw` would just require an if statement (which I'm happy to write myself if needed). Presumably the TPU is also doing Householder QR, so the lists should be accessible in that case too, though I can't say how conveniently.

**Comment by shoyer** [comment]
Contributions would be very welcome here!

NumPy has a scale API surface, we simply haven't had time to fill it all in.

**Comment by hawkinsp** [comment]
Two quick comments:
a) the TPU (i.e., non-CPU or GPU) implementation used by JAX is in the XLA source tree in C++: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/client/lib/qr.h
I was the original author of it; feel very free to modify it to compute `h` and `tau` instead. Or feel free to contribute a Python implementation in the style of the LU decomposition in `lax_linalg.py`.

b) If we return `h` and `tau`, how would that interact with autodiff? The autodiff rules are defined in terms of QR as a primitive operation. Maybe it's easy; I haven't thought about it.

**Comment by alewis** [comment]
With regards to a), I have a pure Jax implementation of Householder QR I'd be happy to contribute, but per [https://github.com/google/jax/issues/2308] a bit more fiddling with LAX primitives and masks will be required first. I'll ping you guys when I've done that and you can see what you think.

I'm afraid I don't have much insight into b).

If you'd like, I also have Jax code to convert from `[h, tau]` to the so-called "WY" representation, which also permits `Q@A` to be computed in ~O(m n) operations, but maintains the matrix-matrix (rather than matrix-vector) fraction of those operations. I will also likely add code for other representations soon. This is not implemented in NumPy, though, so perhaps it's a bridge too far for Jax.
